Added Indentation section to document the exact exapected behavior of…

mipo256 opened the pull request:
As discussed [in this issue](https://github.com/editorconfig/editorconfig/issues/521), there turned out to be a lot of undocumented behavior in terms of indentation style in use. This PR aims to address the problem mentioned in the aforementioned issue by documenting the behavior of indentation options.

CC: @xuhdev 

<!-- readthedocs-preview editorconfig-specification start -->
----
📚 Documentation preview 📚: https://editorconfig-specification--65.org.readthedocs.build/

<!-- readthedocs-preview editorconfig-specification end -->

mipo256 commented:
Hey @xuhdev, any chance to review the PR soon? We'll soon complete the plugin, we're going to need the clarification to justify the behavior of the plugin.

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,52 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+Indentation
+=================================
+The indentation related options (``indent_style``, ``indent_size`` and ``tab_width``) require a special documentation
+section to specify their behavior. Let's assume we have the following code snippet::
```
```suggestion
section to specify their behavior. Consider the following code snippet::
```

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,52 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+        for i in source.split(" "):
+            print(i)
+
+The ``indent_size`` setting for this code snippet equals to 4, because ``indent_size`` actually means how much columns are required
```
```suggestion
The ``indent_size`` setting for this code snippet equals 4, because ``indent_size`` means how many columns are required
```

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,52 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+
+The ``indent_size`` setting for this code snippet equals to 4, because ``indent_size`` actually means how much columns are required
+to indent the next line in relation to previous (if indentation, of course, is applicable for this line). Then the next question
+is *how* this indentation of 4 columns is actually achieved. It may be just 4 consequent spaces/soft tabs, or it equally might
```
```suggestion
is *how* this indentation of 4 columns is achieved. It may be 4 consequent spaces/soft tabs,
```

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,52 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+The ``indent_size`` setting for this code snippet equals to 4, because ``indent_size`` actually means how much columns are required
+to indent the next line in relation to previous (if indentation, of course, is applicable for this line). Then the next question
+is *how* this indentation of 4 columns is actually achieved. It may be just 4 consequent spaces/soft tabs, or it equally might
+be just a single tab with width equal to 4, or two tabs with width equal to 2.
```
```suggestion
a single tab with width equal to 4, or two tabs with width equal to 2.
```

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,52 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+    indent_size = 4
+    tab_width = 3
+
+As we can notice, the ``indent_size`` of 4 is not achievable by placing 1 or 2 consequent tabs, because the ``tab_width``. Therefore,
```
```suggestion
The ``indent_size`` of 4 is not achievable by placing 1 or 2 consequent tabs, because the ``tab_width``. Therefore,
```

mipo256 replied:
Seems I've overlooked the ending of the sentence:

> As we can notice, the ``indent_size`` of 4 is not achievable by placing 1 or 2 consequent tabs, because the ``tab_width``

The ending of the sentence is not clear :) Applying your suggestion, the final result would look like this:

> The ``indent_size`` of 4 is not achievable by placing 1 or 2 consequent tabs, because the ``tab_width = 3``

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,52 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+As we can notice, the ``indent_size`` of 4 is not achievable by placing 1 or 2 consequent tabs, because the ``tab_width``. Therefore,
+in order to comply with this editorconfig configuration, the new lines (where indentation is applicable, of course) **must be precisely
+indented with one tab, and one space**. That is because by placing one tab we're not achieving the ``indent_size`` required, but by
+placing the 2 consequent tabs we're overreaching. Therefore, although the preferable ``indent_style`` is tab, we still have to fallback
```
```suggestion
placing the 2 consequent tabs we're overreaching. Therefore, although ``indent_style`` is ``tab``, we still have to fallback
```

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,52 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+placing the 2 consequent tabs we're overreaching. Therefore, although the preferable ``indent_style`` is tab, we still have to fallback
+to one space character to fulfill the requirement.
+
+Also as a consequence, when we have the following editorconfig rules defined::
```
```suggestion
For another example, if we have the following editorconfig rules defined::
```

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,52 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+    indent_size = 8
+    tab_width = 4
+
+One **MUST** expect that spaces will not be used at all for indentation, since all the indentation can be achieved via tabs only
```
```suggestion
One **MUST** expect that spaces will not be used at all for indentation, since all the indentation can be achieved via tabs only.
```

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,52 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+    [another_file.py]
+    indent_style = tab
+    indent_size = 8
+    tab_width = 4
```
I think it's better to replace this example with one that has `indent_size` < `tab_width`, but I can add the example myself later.

mipo256 replied:
Sounds reasonable. Now, we have these two sentences: 

<pre>
As a side note - it is *not* restricted (although, it does not make much sense) by spec to have ``indent_size`` less than the ``tab_width``.
In this case, regardless of ``indent_style``, only the spaces can be used to achieve the appropriate ``indent_size`` for obvious reasons.
</pre>

We can enrich the example above to look like this:

<pre>
As a side note - it is *not* restricted (although, it does not make much sense) by spec to have ``indent_size`` less than the ``tab_width``:

    [another_file.py]
    indent_style = tab
    indent_size = 4
    tab_width = 8

In this case, regardless of ``indent_style``, only the spaces can be used to achieve the appropriate ``indent_size`` for obvious reasons.
</pre>

@xuhdev What are your thoughts on this?

mipo256 replied:
@xuhdev  Okay, I see, I'll add an example, thx!

mipo256 replied:
Done, please, see commit: https://github.com/editorconfig/specification/pull/65/commits/4343eef1bc1c655f74017d90db7a561d92031edf

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,52 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
 
+.. indentation:
+
+Indentation
```
```suggestion
Indentation (Non-Normative)
```

xuhdev commented on index.rst:
```diff
@@ -235,7 +235,9 @@ and the supported values associated with them:
      - Supported values
    * - ``indent_style``
      - Set to ``tab`` or ``space`` to use hard tabs or soft tabs respectively. The
-       values are case insensitive.
+       values are case insensitive. Note that this settings signifies the indentation
+       style to be **preferred whenever possible**. The exact semantics with examples can be
+       found in the indentation_ section
```
How about changing the whole bullet point to:

```
Set to ``tab`` or ``space`` to use tabs or spaces for indentation, respectively. ``tab`` implies that an indentation is to be filled by as many hard tabs as possible, with the rest of the indentation filled by spaces. A non-normative explanation can be
found in the indentation_ section. The values are case insensitive.
```

mipo256 replied:
Sounds great, I'll just change the beginning of the seconds sentence to ``tab``.

xuhdev reviewed (commented):
Sorry for the delay and thanks for the writeup!

I've made a suggestion to rewrite the bullet point, and some grammatical changes in the indentation section.

mipo256 commented:
Hey @xuhdev!
Thanks for suggestions, I've applied almost all of them. There are just a couple of cases where I proposed changes based on your suggestions as well. Please, take a look at it :)

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,52 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+.. indentation:
+
+Indentation (Non-Normative)
+=================================
```
```suggestion
========================
```

mipo256 commented:
Hey @xuhdev!
Thanks for the review. I've addressed the remaining issues in the following commit: https://github.com/editorconfig/specification/pull/65/commits/4343eef1bc1c655f74017d90db7a561d92031edf

P.S: Once we're okay with the final state of the PR, I'll squash the commits into a single one, now I'm retaining history for the ease of review.

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,66 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+            return True
+
+In this case, the line where the ``if`` statement condition is specified is indented with 4 spaces, because the ``indent_size = 4``
+and the tab cannot fit in. On the other hand, the line with ``return`` statement must be indented with the single tab, because the
```
```suggestion
and the tab cannot fit in. On the other hand, the line with ``return`` statement must be indented with one tab, because the
```

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,66 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+
+In this case, the line where the ``if`` statement condition is specified is indented with 4 spaces, because the ``indent_size = 4``
+and the tab cannot fit in. On the other hand, the line with ``return`` statement must be indented with the single tab, because the
+indentation level for this line is 8 columns, and tag can fit it.
```
```suggestion
indentation level for this line is 8 columns, and a tab can fit in.
```

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,66 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+    tab_width = 3
+
+The ``indent_size`` of 4 is not achievable by placing 1 or 2 consequent tabs, because ``tab_width = 3``. Therefore,
+in order to comply with this editorconfig configuration, the new lines (where indentation is applicable, of course) **must be precisely
```
```suggestion
in order to comply with this EditorConfig configuration, the new lines (where indentation is applicable) **must be precisely
```

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,66 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+The ``indent_size`` of 4 is not achievable by placing 1 or 2 consequent tabs, because ``tab_width = 3``. Therefore,
+in order to comply with this editorconfig configuration, the new lines (where indentation is applicable, of course) **must be precisely
+indented with one tab, and one space**. That is because by placing one tab we're not achieving the ``indent_size`` required, but by
+placing the 2 consequent tabs we're overreaching. Therefore, although ``indent_style`` is ``tab``, we still have to fallback
+to one space character to fulfill the requirement.
```
```suggestion
placing the 2 consequent tabs we're overreaching. Therefore, although ``indent_style`` is ``tab``, we still have to supplement
with one space character to fulfill the requirement.
```

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,66 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+placing the 2 consequent tabs we're overreaching. Therefore, although ``indent_style`` is ``tab``, we still have to fallback
+to one space character to fulfill the requirement.
+
+For another example, if we have the following editorconfig rules defined::
```
```suggestion
For another example, if we have the following EditorConfig rules defined:

.. code-block:: ini
```

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,66 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+
+One **MUST** expect that spaces will not be used at all for indentation, since all the indentation can be achieved via tabs only.
+
+As a side note - it is possible to have ``indent_size`` less then the ``tab_width``.
```
```suggestion
Additionally, it is possible to have ``indent_size`` less then the ``tab_width``.
```

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,66 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+    indent_size = 4
+    tab_width = 8
+
+To understand the way it works, let's look at the following example:
```
```suggestion
To understand the way it works, let's look at the following example:

.. code-block:: python
```

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,66 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+Indentation (Non-Normative)
+===========================
+The indentation related options (``indent_style``, ``indent_size`` and ``tab_width``) require a special documentation
+section to specify their behavior. Consider the following code snippet::
```
```suggestion
section to specify their behavior. Consider the following code snippet:

.. code-block:: python
```

xuhdev commented on index.rst:
```diff
@@ -285,6 +287,66 @@ Pair keys are case insensitive. All keys are lowercased after parsing.
+
+This is when ``indent_style`` comes into picture. It specifies what character should be used **whenever possible** in order to
+achieve the indentation size specified in ``indent_size``. To fully understand what "whenever possible" actually means, lets
+assume that the editorconfig rules are specified for the file above::
```
```suggestion
assume that the editorconfig rules are specified for the file above:

.. code-block:: ini
```

xuhdev reviewed (commented):
Looks good to me now! I added some comments for minor language and format changes. Other than those, the PR is good to go.

mipo256 commented:
Hey @xuhdev!
I agree with all of yours suggestions, I've fixed them, can you re-review?) I'll squash the commits once you'll approve

xuhdev reviewed (approved):
Perfect! GitHub can also squash merge, do you want me go ahead, or you have something else in mind?

mipo256 commented:
No, everything's good. I've squashed the commits to ease the job for you :smile:  You can proceed with merging it into the master trunk :smile
